Remotion/Core/Tools/ToolRunner.cs

```diff
@@ -0,0 +1,70 @@
+﻿using System;
+using System.Configuration;
+using System.IO;
+using System.Reflection;
+using JetBrains.Annotations;
+using Remotion.Utilities;
+
+#if NET
+namespace Remotion.Tools
+{
+  /// <summary>
+  /// Base class for executing code that should run in a different base directory and or use a different config file.
+  /// </summary>
+  public abstract class ToolRunner
+  {
+    private readonly string _baseDirectory;
+    [CanBeNull]
+    private readonly string _configFile;
+
+    public ToolRunner (string baseDirectory, [CanBeNull] string configFile)
+    {
+      ArgumentUtility.CheckNotNull("baseDirectory", baseDirectory);
+
+      _baseDirectory = baseDirectory;
+      _configFile = configFile;
+    }
+
+    public void Run ()
+    {
+      var baseDirectory = Path.GetFullPath(_baseDirectory);
+      AppDomain.CurrentDomain.SetData("APP_CONTEXT_BASE_DIRECTORY", baseDirectory);
+
+      if (!string.IsNullOrEmpty(_configFile))
+      {
+        var configFile = Path.GetFullPath(_configFile);
+        if (!File.Exists(configFile))
+        {
+          throw new FileNotFoundException(
+              string.Format(
+                  "The configuration file supplied by the 'config' parameter was not found.\r\nFile: {0}",
+                  configFile),
+              configFile);
+        }
+
+        var configurationState = (int)typeof(ConfigurationManager).GetField("s_initState", BindingFlags.NonPublic | BindingFlags.Static)!.GetValue(null)!;
+        if (configurationState != 0)
+          throw new InvalidOperationException("Cannot set the active config file as the configuration is already loaded.");
+
+        AppDomain.CurrentDomain.SetData("APP_CONFIG_FILE", configFile);
+      }
+
+      // Changing the base directory won't load assemblies from there so we need to resolve them manually
+      AppDomain.CurrentDomain.AssemblyResolve += (_, eventArgs) =>
+      {
+        var assemblyName = new AssemblyName(eventArgs.Name);
+
+        var assemblyFilePath = Path.Combine(baseDirectory, assemblyName.Name + ".dll");
+        if (File.Exists(assemblyFilePath))
+          return Assembly.LoadFile(assemblyFilePath);
+
+        return null;
+      };
+
+      CallbackHandler();
```

Comment:
I would prefer a better name. It's no longer an AppDomainCallbackHandler.

Question: since it's totally different anyway, should we drop the basetype-concept and just provide a helper method for intiializing the AppDomain that's then called before doing the actual implementation, i.e.:
```
RdbmsToolsRunner
{
   public void Run()
   {
    AppDomainSetupHelper.Intiialize(basedir, configfile);
    // do stuff previously located in Callbackhandler()
    }
}
```

I'm aware that this changes the contract, so it may not be the right time to do this. It's just a thought. If it's not a good idea at this time, we should leave a comment to consider updating the API when we drop .net framework.

On the other hand, we could also simply add a try-finally around the CallbackHandler and re-register the AssemblyResolve once we're done, reset the base-directoy and clear the config.

Reply:
I don't have a better name yet, just a general idea: the concept of "callback handler" isn't true for .NET, so we need something else. This should servce your intention of highlighting the difference. The new API will be meant to implement the business logic, just like before, but it's no longer an AppDomain callback, just simply executing our code after the setup. My non-creative approach would be `RunImplementation` or `ExecuteWithNewBaseDirectory` / `RunInNewBaseDirectory` or something of the sort.